### PR TITLE
[v6r20] fix task queues selection at matching time

### PR DESCRIPTION
**Pull request by fstagni**
Mostly fixes for TaskQueueDB.py. Please have a look and comment.

BEGINRELEASENOTES

*WMS
CHANGE: TQ matching (TaskQueueDB.py): when "ANY" is specified, don't exclude task queues (fix with "Platforms" matching in mind)
CHANGE: SiteDirector: split method getPlatforms, for extension purposes 

ENDRELEASENOTES

### CODE REVIEW

File: tests/Integration/WorkloadManagementSystem/Test_TaskQueueDB.py [hunk 1]
```diff
@@ -1,152 +1,355 @@
...
+  assert int(result['Value'][0][0]) in [tq_job4, tq_job6, tq_job7]
+  assert len(result['Value']) == 2
+
+  # new platform appears
+  # centos8 (> centos7)
+  result = tqDB.matchAndGetTaskQueue({'Setup': 'aSetup', 'CPUTime': 50000,
+                                      'Platform': 'centos8'},
+                                     numQueuesToGet=5)
+  assert result['OK'] is True
+  # FIXME: here, I would like to see 3 TQs matched: those for slc6 + centos7 + ANY
```
Comment (fstagni) [inline tests/Integration/WorkloadManagementSystem/Test_TaskQueueDB.py]: Let me know what you think of this

**Comment by fstagni** [comment]
I added WIP, as I am doing few more tests. Review anyway welcome!

**Comment by fstagni** [comment]
@atsareg please have a look at this one.
Read also the test file that I expanded, there are some "FIXME" comments that you may want to look into.

File: tests/Integration/WorkloadManagementSystem/Test_TaskQueueDB.py (hunk 1 above)
Comment (atsareg) [reply tests/Integration/WorkloadManagementSystem/Test_TaskQueueDB.py]: I do not understand the relation centos8 > centos7. Do you mean that on centos8 platform jobs requiring centos7 can be executed ? Where this rlation is formally specified ? in OSCompatibility settings ? This looks very confusing to me.

**Comment by atsareg** [comment]
I am trying to understand what do you want to achieve with the changes in this PR. Especially the need for ANY value for the Platform and Tag. What is an example use case that is not covered in the present matching mechanism ?
We discussed the possibility to drop OSCompatibility mechanism which is there for mostly historical reasons. The suggestion was to just drop it and then:

1. CEs/Queues must have a Platform parameter which is expressed in terms of the DIRAC nomenclature, e.g. Linux_x86_64_glibc-2.17 (for centos7). This is a single value, not a list.
2. CEs/Queues can define other parameters as Tags, e.g. compiler options or processor instructions
3. Platform and Tags are used when matching TaskQueues either from the SiteDirector (and then configuration settings are used) or from a pilot (then the Platform and Tags can be discovered on the worker node).
4. Jobs, if they care, can specify Platform. Jobs can specify several Platforms, in this case the logic is that for matching the resource (queue) should have Platform value one in the requested list. Note that this is different from the Tags. ANY Platform is not foreseen, simply because if a job does not care it is not specifying it at all
5. Finally, jobs can specify any number of Tags. If Tags are specified, ALL of them must be satisfied by the CE/Queue resource description. This is different compared to the Platform parameter. Also ANY value for Tags makes no sense.

The idea is then to go through the Configuration Resources section and add Platform to all the queues description. The BDII2CS agent should also be updated to set the appropriate Platform (OSCompatibility can be kept for this purpose and this purpose only). Once this is done, we can drop the OSCompatibility in the matching.
If you want to have a logic that some platform can be more generic than others and can "contain" others, then I would really like to see the use case for that. To me this looks like very confusing.

**Comment by fstagni** [comment]
In this PR there are 2 things:

1. trying to fix what is IMHO a bug in TQs selection where there's a mix of TQs with platforms and of TQs without platforms
2. adding tests for tags, and this is mostly for my personal understanding. 

Regarding the first point, I did the following test on LHCb production setup, where there's a mix of TQs with platforms and without platform. I consider that TQs without a platform specified (or with "ANY" platform) should be matched always, because it means that they can run everywhere.

```
from DIRAC.Core.Base.Script import parseCommandLine
parseCommandLine()

from DIRAC import gLogger
from DIRAC.Core.DISET.RPCClient import RPCClient

gLogger.setLevel('DEBUG')


tqDictNoPlatforms = {'CPUTime': 9999999,
'Community': 'lhcb',
'OwnerGroup': ['lhcb_user'],
'Setup': 'LHCb-Production',
'Site': ['LCG.CERN.cern', 'LCG.CNAF.it'],
'SubmitPool': ''}

tqDictAnyPlatforms = {'CPUTime': 9999999,
'Community': 'lhcb',
'OwnerGroup': ['lhcb_user'],
'Setup': 'LHCb-Production',
'Site': ['LCG.CERN.cern', 'LCG.CNAF.it'],
'Platform': 'any',
'SubmitPool': ''}


tqDict = {'CPUTime': 9999999,
'Community': 'lhcb',
'OwnerGroup': ['lhcb_user'],
'Platform': ['x86_64_ScientificLinux_6.1_Carbon',
'x86_64_CentOS_Core_7.4',
'x86_64-slc6',
'x86_64-slc5',
'x86_64_ScientificLinuxCERNSLC_6.3_Carbon',
... long list
'x86_64_CentOS_6.9_Final',
'x86_64_CentOS_Final_6.4',
'x86_64_CentOS_Final_6.7',
'x86_64_CentOS_Final_6.8',
'x86_64_CentOS_Final_6.9',
'x86_64-slc5-gcc41-opt',
'x86_64-slc5-gcc43-opt',
'x86_64-slc5-gcc46-opt',
'x86_64-slc6-gcc44-opt',
'x86_64-slc6-gcc46-opt',
'x86_64-slc6-gcc48-opt',
'x86_64-slc6-gcc49-opt'],
'Setup': 'LHCb-Production',
'Site': ['LCG.CERN.cern', 'LCG.CNAF.it'],
'SubmitPool': ''}


tqDictX = {'CPUTime': 9999999,
'Community': 'lhcb',
'OwnerGroup': ['lhcb_user'],
'Setup': 'LHCb-Production',
'Site': ['LCG.CERN.cern', 'LCG.CNAF.it'],
'Platform': ['any','x86_64-slc6','x86_64-slc5'],
'SubmitPool': ''}




rpcMatcher = RPCClient( "WorkloadManagement/Matcher" )

resForTQNoPlatfroms = rpcMatcher.getMatchingTaskQueues( tqDictNoPlatforms )
```

Results:
#### tqDict and tqDictX give the same result (long list of TQs, those with a platform specified)
#### but tqDictAnyPlatforms and tqDictNoPlatforms give a much shorter list, which correspond to ONLY the TQ that don't have a platform (they are not many) 

This is because today the matching is **strict**, and IMHO this is a bug.

**Comment by atsareg** [comment]
I do not see inconsistency. If a TQ does specify a Platform, this is a requirement and must be explicitly satisfied by the resource. So, if a resource does not specify a Platform, TQ's requiring a particular platform will not be selected. This is a correct behavior.
On the other hand in your example I do not see a use case for a resource to present multiple Platforms, especially non-compatible ones.

**Comment by fstagni** [comment]
On the other points of your comment, I would first like to remind that these platforms (and the TQs matching in general) are used today in 2 moments:

1. from the JobAgent, when it's time to match a job, given the _node_ (or _slot_) description: we often call it `ceDict`, but in fact it's really the description of the capabilities of the node where the JobAgent is running, at least for the inProcessCE case, which is the vast majority of the cases. 
2. from the SiteDirector when deciding if to send a pilot or not, and in this case the `ceDict` is really the description of the CE that is found in the CS

The first match (from the JobAgent) has to be a precise match. The second one (from the SiteDirector) does not need to be precise. So, when we are talking about dropping the OSCompatibility mechanism, we are talking about dropping it for the case of the SiteDirector, and this PR addresses this case with the small changes I have done in the SiteDirector.

**Comment by fstagni** [comment]
Removed the "WIP". Please review carefully, and answer to the "FIXME" questions in the integration test.

**Comment by fstagni** [comment]
Removed the "WIP". Please review carefully, and answer to the "FIXME" questions in the integration test.

**Comment by fstagni** [comment]
Removed the "WIP". Please review carefully, and answer to the "FIXME" questions in the integration test.

**Comment by fstagni** [comment]
Removed the "WIP". Please review carefully, and answer to the "FIXME" questions in the integration test.

**Comment by fstagni** [comment]
I think I implemented what we discussed. Please have a look @atsareg

**Comment by fstagni** [comment]
Hi @atsareg please have one more look here. Honestly, after several trials, I ended up with the code solution. I didn't restore the "strict matching" everywhere, because it makes the query very long, which I am afraid impacts performances.

File: WorkloadManagementSystem/DB/TaskQueueDB.py [hunk 2]
```diff
@@ -678,44 +704,67 @@ def __generateTQMatchSQL(self, tqMatchDict, numQueuesToGet=1, negativeCond=None)
...
+          # Is there something to consider?
+          if not tqMatchDict.get(field) \
+             or \
+             isinstance(tqMatchDict.get(field), str) \
+             and tqMatchDict.get(field).lower() in starValues \
+             or \
+             isinstance(tqMatchDict.get(field), list) \
+             and any(x in [fv.lower() for fv in tqMatchDict.get(field)] for x in starValues):
+            continue
+          print [fv.lower() for fv in tqMatchDict.get(field)]
```
Comment (atsareg) [inline WorkloadManagementSystem/DB/TaskQueueDB.py]: Debugging to go ?

Comment (fstagni) [reply WorkloadManagementSystem/DB/TaskQueueDB.py]: Removed

**Pull request merged**